Fix default processing instructions for INTER

gemmaguest opened the pull request:
**Report to:** Max at ISIS

**Note** This PR contains a parameter file change that will take immediate effect when merged. It's fine for this to be merged together with the code - the current parameter file values are useless anyway due to the bug in question.

**Description**

This PR fixes a problem where the default processing instructions were being constructed incorrectly, causing confusing reduction failures. It also updates the INTER parameters file to set more sensible defaults. The main changes are:
- Change `ReflectometryWorkflowBase::findProcessingInstructions` to construct the processing instructions using separator `-` instead of `:`; the latter is incorrect because it causes spectra to be grouped rather than summed, meaning the reduction fails.
- Change the same function to support a `MultiDetectorStop` default in the  parameters file. This is required because the INTER linear detector no longer has a valid detector at its final histogram so we need to clip this out.
- Update the INTER parameters file to set more sensible start/stop for the multi detector. Max suggested the spectra `70-90`. The file takes workspace indices for which the equivalents are `69-89`.

This PR also does the following minor tidying:
- Remove test files from the parent directory CMakeLists.txt because they are included in the test subdirectory CMakeLists.txt
- Minor refactoring to move declaring the `AnalysisMode` property into a function in `ReflectometryWorkflowBase` to reduce duplication and make it so it can be initialised in the tests.

**To test:**

Run this script, which tests a few different reductions and compares the results when using the defaults vs the results when specifying the same processing instructions in the algorithm call.
```
# test summing in Q for multi detector - default spectra are 70-90
Load(Filename='INTER45455', OutputWorkspace='INTER45455')
qbin1, q1 = ReflectometryReductionOneAuto(InputWorkspace='INTER45455', SummationType='SumInQ', ReductionType='DivergentBeam', AnalysisMode='MultiDetectorAnalysis', ThetaIn=0.5)
qbin2, q2 = ReflectometryReductionOneAuto(InputWorkspace='INTER45455', SummationType='SumInQ', ReductionType='DivergentBeam', AnalysisMode='MultiDetectorAnalysis', ThetaIn=0.5, ProcessingInstructions='70-90')
CompareWorkspaces(Workspace1='qbin1', Workspace2='qbin2', CheckSample=True, ToleranceRelErr=True, CheckAllData=True)

# test summing in lambda for multi detector - default spectra are 70-90
qbin3, q3 = ReflectometryReductionOneAuto(InputWorkspace='INTER45455', AnalysisMode='MultiDetectorAnalysis', ThetaIn=0.5)
qbin4, q4 = ReflectometryReductionOneAuto(InputWorkspace='INTER45455', AnalysisMode='MultiDetectorAnalysis', ThetaIn=0.5, ProcessingInstructions='70-90')
CompareWorkspaces(Workspace1='qbin3', Workspace2='qbin4', CheckSample=True, ToleranceRelErr=True, CheckAllData=True)

# Test point detector - default spectrum is 4
Load(Filename='INTER13460', OutputWorkspace='INTER13460')
qbin5, q5 = ReflectometryReductionOneAuto(InputWorkspace='INTER13460', AnalysisMode='PointDetectorAnalysis', ThetaIn=0.5)
qbin6, q6 = ReflectometryReductionOneAuto(InputWorkspace='INTER13460', AnalysisMode='PointDetectorAnalysis', ThetaIn=0.5, ProcessingInstructions='4')
CompareWorkspaces(Workspace1='qbin5', Workspace2='qbin6', CheckSample=True, ToleranceRelErr=True, CheckAllData=True)
```

Ensure the log shows these three messages, meaning that we got the same results with the defaults as the explicit calls:
```
The workspaces "qbin1" and "qbin2" matched!
The workspaces "qbin3" and "qbin4" matched!
The workspaces "qbin5" and "qbin6" matched!
```

Fixes #33640

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

gemmaguest committed:
Add ReflectometryWorkflowBase tests and minor refactoring

- Add a test file for ReflectometryWorkflowBase2. This is an abstract class but has a lot of untested functionality that is not tested in the derived classes so I've added the file so we can start adding tests for these functions. Due to it being abstract with protected members I've created a stub class derived from it which is a friend of the tests.
- I've started by adding tests for findProcessingInstructions due to upcoming work planning to change this.
- I've done some minor refactoring to move duplicated code that declares the AnalysisMode property into the base class because this is required by the tests.

Note that I have not been able to add tests for the invalid cases where the function throws if properties are missing from the parameters file. This is because of a quirk in Mantid where the previously-loaded parameters file for the same instrument is loaded when you call LoadInstrument, i.e. it then starts with populated parameters for the ones I am trying to test that are empty, and even when I re-load the correct parameters file (where they are missing) they remain populated. It is beyond the scope of this commit to fix this at this point.

Re #33640

gemmaguest committed:
Remove duplicate cmakelist entries

These are in the test subdirectory CMakeLists.txt file so are not required here

gemmaguest committed:
Fix incorrect default processing instructions

This commit fixes a couple of issues contributing to the default processing instructions being incorrect:
- The processing instructions were constructed with a ":" instead of a "-" separator meaning they would be grouped rather than summed, causing the reduction to fail
- The multi detector did not allow for an end index to be set; INTER currently has an additional invalid spectrum which needs to be clipped out, hence does not work if we include all spectra to the end

Re #33640

gemmaguest committed:
Update INTER parameters file with default processing instructions

The currently-specified range of index 6 to the end is not sensible and causes the reduction to fail because it includes an invalid spectrum at the max index of the workspace. A more sensible default range for the INTER linear detecor is spectra 70-90 (workspace indices 69-89)

Re #33640

gemmaguest committed:
Add release note for the processing instructions fix

Re #33640

DavidFair reviewed (commented):
This fixes the issue for me locally
Just a couple of questions / points:

gemmaguest commented on Framework/Reflectometry/src/ReflectometryWorkflowBase2.cpp:
```diff
@@ -685,31 +696,33 @@ std::string ReflectometryWorkflowBase2::findProcessingInstructions(const Instrum
-        std::to_string(static_cast<int>(multiStart[0])) + ":" + std::to_string(inputWS->getNumberHistograms() - 1);
+    }
+    // Use last workspace index if stop is empty
+    auto const stop = multiStop.empty() ? inputWS->getNumberHistograms() - 1 : static_cast<int>(multiStop[0]);
```
Good point, there's no checking on this at all - I'll add verification for all of them

gemmaguest committed:
Add validation for detector start/stop values in IPF

Check that values are within the workspace index range.
Minor refactoring to reduce duplication.

Re #33640

gemmaguest commented on docs/source/release/v6.4.0/Reflectometry/Bugfixes/33648.rst:
```diff
@@ -0,0 +1 @@
+- Fix a problem where the default processing instructions were incorrect causing confusing reduction errors
```
Good shout. I'd forgotten what the original (really confusing!) error message was :sweat_smile:

gemmaguest replied:
I'll be managing the patch candidate so thanks for the reminder :+1:

robertapplin commented on Framework/Reflectometry/src/ReflectometryWorkflowBase2.cpp:
```diff
@@ -681,37 +722,34 @@ std::string ReflectometryWorkflowBase2::findProcessingInstructions(const Instrum
+    maybeStart = getDetectorParamOrNone(instrument, inputWS, "MultiDetectorStart");
+    maybeStop = getDetectorParamOrNone(instrument, inputWS, "MultiDetectorStop");
+    if (!maybeStart) {
+      throw std::runtime_error("Could not find 'MultiDetectorStart in parameter file. Please provide processing "
```
A think there is a missing single quotation in this line after `MultiDetectorStart`

gemmaguest replied:
Good spot!

robertapplin reviewed (changes requested):
Code looks good, and some nice tests have been added. The testing instructions also work as described. I just spotted one small typo, and then had a question which is for my own benefit

robertapplin commented on Framework/Reflectometry/src/ReflectometryWorkflowBase2.cpp:
```diff
@@ -681,37 +722,34 @@ std::string ReflectometryWorkflowBase2::findProcessingInstructions(const Instrum
+    }
+    if (!maybeStop) {
+      // Default to the last workspace index if stop is not given
+      maybeStop = inputWS->getNumberHistograms() - 1;
```
Also just want to check that we don't expect `getNumberHistograms()` to return 0?

gemmaguest replied:
Ah, good point, we don't but I should throw or assert here. I'll have a think about what's best.

gemmaguest replied:
I've put in an assert for now because we really shouldn't get here with 0 histograms. It should probably be validated earlier on in validateInputs but that's beyond the scope of this PR. I'll create an issue for it.

gemmaguest commented on Framework/Reflectometry/src/ReflectometryWorkflowBase2.cpp:
```diff
@@ -83,10 +83,51 @@ std::vector<size_t> getProcessingInstructionsAsIndices(std::string const &instru
+
+  auto const wsIndex = static_cast<size_t>(value[0]);
+  if (wsIndex >= inputWS->getNumberHistograms()) {
+    std::ostringstream msg;
+    msg << "Parameter file value " << param << "=" << wsIndex
+        << " is out of range; max workspace index= " << inputWS->getNumberHistograms() - 1 << ")";
+    throw std::out_of_range(msg.str());
```
If anything it's probably less efficient but won't make much difference in this case. I use it just because it automatically converts the numbers so avoids having to do the `std::to_string` calls. I just think it makes it a bit more readable when you've got a lot of arguments.

robertapplin replied:
Ah ok, that makes sense - thanks!

gemmaguest commented:
@robertapplin This is ready for another look

robertapplin reviewed (approved):
Thanks for those changes, this is looking good to go 👍
